### ISSUE TITLE
many relationship

### ISSUE THREAD

**Issue by emsici** [thedevdojo/voyager#706]
- Laravel Version: 5.3
- Voyager Version:  (v0.10.13)
- PHP Version: PHP 5.6.4-4ubuntu6.4 (cli) (built: Oct 28 2015 01:21:29)
- Database Driver & Version: mysql  Ver 15.1 Distrib 10.0.23-MariaDB, for debian-linux-gnu (x86_64) using readline 5.2

### Description:
Tables: certifications - id, name, description
clients - certifications_id (multiple select), name, etc

I have created a BREAD for this, but i can't add or edit multiple values from certifications table.

The BREAD is {
    "relationship": {
        "key": "id",
        "label": "name"
    }
}

Client.php

> class Client extends Model {
>   protected $table = 'clients';
>   
> 
>   public function certifications_id() { 
>     return $this->belongsTo(Certification::class);
>   }
> 
> }

mySQL:
**clients:**
> id	int(10) unsigned Auto Increment	 
> type	varchar(255)	 
> name	varchar(255)	 
> created_at	timestamp NULL	 
> updated_at	timestamp NULL

**certifications:**
id	int(10) unsigned Auto Increment	 
name	varchar(255)	 
indicativ	varchar(255)	 
certifications_id	varchar(255)	 
created_at	timestamp NULL	 
updated_at	timestamp NULL
### Steps To Reproduce:
![snip](https://cloud.githubusercontent.com/assets/7910648/22807534/2e00880a-ef30-11e6-8c81-7b71b0a51ce9.PNG)

![snip2](https://cloud.githubusercontent.com/assets/7910648/22807533/2dff9d0a-ef30-11e6-8866-58ca20992be7.PNG)

**Comment by emsici** [comment]
There will be a fix in next release or can i do something to fix this?

**Comment by Frondor** [comment]
Remember than after #638 you have to name your relationship methods in `camelCase`, not `snake_case` anymore. I wonder how your code goes that far, it should fail to resolve the relationship if you're using 0.10.13.

**Comment by emsici** [comment]
Relations are working, but i can't insert to mysql because it says empty value, i have camelCase relations at this moment.

**Comment by Frondor** [comment]
I've seen this error once but I use validation for each field. The ones that you can allow to be null, simply uses the `nullable` rule. Hence, the `required` ones need that one rule.
But to be honest, I can't reproduce it right now at 0.10.13, I don't know about @marktopper 

EDIT: If that field can be `null`, then you have to specify `->nullable()` in your migration file, and then write the validation rules to allow null values. If that's not the case, then you also need to write validation logic and make it `required`

**Comment by marktopper** [comment]
Haven't been able to reproduce this either...

**Comment by emsici** [comment]
![1](https://cloud.githubusercontent.com/assets/7910648/22856222/267521e2-f096-11e6-8535-7a9bc68c6590.PNG)
![2](https://cloud.githubusercontent.com/assets/7910648/22856223/2676b6ba-f096-11e6-868d-b471ce25d520.PNG)
![3](https://cloud.githubusercontent.com/assets/7910648/22856221/2671c1fa-f096-11e6-84a9-88938ca00363.PNG)
![4](https://cloud.githubusercontent.com/assets/7910648/22856220/2670947e-f096-11e6-99a2-42af0830a76f.PNG)
![5](https://cloud.githubusercontent.com/assets/7910648/22856219/26708916-f096-11e6-9836-0a0fd4809fd6.PNG)
![6](https://cloud.githubusercontent.com/assets/7910648/22856294/1fc1a982-f097-11e6-9121-4cd35231038a.PNG)

if i use nullable() > Call to undefined method Illuminate\Database\Query\Builder::sync()
when adding

**Comment by Frondor** [comment]
Now that's a new error.. Can you edit https://github.com/the-control-group/voyager/blob/master/resources/views/bread/browse.blade.php#L40 and use on the previous line `dd($data->{$row->field})` to see what value you're getting?

**Comment by emsici** [comment]
![1](https://cloud.githubusercontent.com/assets/7910648/22857154/69e1892c-f0a8-11e6-9dfb-eba02aa0a991.PNG)
![2](https://cloud.githubusercontent.com/assets/7910648/22857153/69e0e3e6-f0a8-11e6-88ba-898375d3d5cd.PNG)
![3](https://cloud.githubusercontent.com/assets/7910648/22857152/69e0b632-f0a8-11e6-89e4-4e34ebfad30c.PNG)
![4](https://cloud.githubusercontent.com/assets/7910648/22857151/69e00cbe-f0a8-11e6-9369-29cfa55cd810.PNG)
![5](https://cloud.githubusercontent.com/assets/7910648/22857150/69dff936-f0a8-11e6-9870-8e0505f58428.PNG)
![6](https://cloud.githubusercontent.com/assets/7910648/22857155/69e22788-f0a8-11e6-8ac2-5e389d06e529.PNG)
![7](https://cloud.githubusercontent.com/assets/7910648/22857158/69fa76b2-f0a8-11e6-9b77-3b39126ae2a6.PNG)
![8](https://cloud.githubusercontent.com/assets/7910648/22857157/69f8f2b0-f0a8-11e6-9eca-d0173a139fae.PNG)
![9](https://cloud.githubusercontent.com/assets/7910648/22857156/69f8ef7c-f0a8-11e6-80a6-2f3bf9a67f37.PNG)

In {{dd($data->{$row->field})}}
resources/views/vendor/voyager/client/browse.blade.php? or where? Yes i can, but give me more info, please.

Made that column and bread from scratch and still no success.

Got Slack or something like that?

**Comment by Frondor** [comment]
mmm.... to be honest, I think that's working as intended. You're supposed to use validation anyway.
I'm already addressing other things, so at the moment my mind can't handle this as well :D

**Comment by marktopper** [comment]
Just a note, it should properly be `public function certificationsId()`.

**Comment by emsici** [comment]
Look where i add a new client, it says it needs to be certifications_id.

Validate what? i have no data like error says :)

**Comment by marktopper** [comment]
@emsici: I see, can you please try to update to version `0.10.14` and test it again?

**Comment by Frondor** [comment]
I can finally reproduce this. @marktopper 

Happens to me with a `image` field which hasn't the current DB value (when editing).
The constraint is fine, those fields can not be null, I think the problem is when voyager prepares the model to be saved, It seems to be ignoring database values, and if they weren't included in the request, they are getting overridden to `NULL`.

I think this is already fixed for `select_multiple` and `select_dropdown` fields, since they already pick up the actual value from DB (when editing), unlike `image` field. But it isn't here where we have to address the error in my opinion;
> I think the problem is when voyager prepares the model to be saved, It seems to be ignoring database values, and if they weren't included in the request, they are getting overridden to `NULL`.

So I have 3 ideas atm

1. We need to revamp the `image` field input to work directly with media manager and save strings instead of files (let the media manager do that).
2. Check which other field types are triggering this error.
3. This error has nothing to do with many to many relationships I believe, so... Let's set it to something like `SQL Error: Field doesn't have a default value` so we may get more feedback

> EDIT: Now I think this is more a developer error than voyager itself. It won't happen at all if a default value (if it uses one) is provided at the moment of migrating the schema to DB, or a proper validation is set up using `details` JSON API.
> Also adding the proper `$fillable` property to the models ofc.

Note to @emsici remember you're saving an array, so you have to cast the attribute as array in your model `public $casts = ['certifications_id' => 'array']`

**Comment by marktopper** [comment]
Great job reproducing this @Frondor. 👍

**Comment by denbatte** [comment]
Have the same issue on a multiple_select (sponsors).
These sponsors have a name, a website and a logo (image).

![image](https://cloud.githubusercontent.com/assets/505196/22993812/ade16f74-f3c4-11e6-91db-dd3bcc96b2e8.png)

![image](https://cloud.githubusercontent.com/assets/505196/22993844/c33a2afa-f3c4-11e6-9269-9fe1d24b478e.png)

**Comment by Frondor** [comment]
@denbatte can you paste your json `details` in `sponsors` field?

**Comment by denbatte** [comment]
This is the JSON in the sponsors BREAD field.
```JSON
{
    "relationship": {
        "key": "id",
        "label": "name"
    }
}
```

Also the model:
```PHP
class Stage extends Model
{
    public function sponsors(){
        return $this->belongsToMany(Sponsor::class, "stages_sponsors");
    }
}
```

**Comment by Frondor** [comment]
Sorry, and the model for that BREAD as well... if you're using one. It would probably be at `app\Stage.php`.
I think there is something with eloquent's `$fillable` property.

@denbatte https://laravel-voyager.slack.com/

**Comment by denbatte** [comment]
Sorry, has been a while since I was active on GitHub. Updated my last post with code sample.
Isn't here a more convenient channel to communicate?

**Comment by denbatte** [comment]
@Frondor do I need an invite for the slack channel?

**Comment by denbatte** [comment]
Ok, did some more testing.
I created a oneToOne relation of a group and a stage and that works perfectly.
Also, I made the sponsors field in the stages table nullable and this made the last error go away.
There is a different error now though. 

![image](https://cloud.githubusercontent.com/assets/505196/22996589/9bbb01a6-f3cf-11e6-9db4-b87b1fd6a622.png)

Having a look at the relationToLink method at the moment.

**Comment by denbatte** [comment]
@Frondor @marktopper I have found the issue. There is a mistake in the relationToLink method.
It uses an array to get the label values in the code but actually, at the moment its a collection.

https://github.com/the-control-group/voyager/blob/master/src/Http/Controllers/Traits/BreadRelationshipParser.php
```PHP
                if (!is_object($item[$field])) {
                    $item[$field] = $relation[$relationData->label]; // HERE IS MISTAKE
                } else {
                    $tmp = $item[$field];
                    $item[$field] = $tmp;
                }
```


I will try to create a PR if I find some time to fix it.

**Comment by Frondor** [comment]
I don't think that's the main problem. This is not only about relationships, since this same error happened with `$timestamps` fields... But that can be another bug. At the moment, it's working fine in my `v0.11` branch

**Comment by emsici** [comment]
Do we need an invite to that Slack channel, or what?

**Comment by denbatte** [comment]
/offtopic
@emsici there is a link with a slack invite website on the Voyager homepage ;)
Just found out.

**Comment by marktopper** [comment]
/offtopic [Get Slack Invite](https://voyager-slack-invitation.herokuapp.com/)

**Comment by tollanes** [comment]
I seem to get the same issue, but who knows if this is just due to my part.
![Image of error](https://puu.sh/u7k8C/a75dd8d371.png)
Here is my parent and pivot table:

```PHP
public function up()
    {
        Schema::create('orders', function (Blueprint $table) {
            $table->increments('id');
            $table->integer('user_id');
            $table->text('comment')->nullable();
            $table->integer('price');
            $table->boolean('sent');
            $table->boolean('paid');
            $table->string('products');
            $table->timestamps();
        });

        Schema::create('order_product', function (Blueprint $table) {
            $table->primary(array('order_id', 'product_id'));
            $table->integer('order_id')->unsigned();
            $table->foreign('order_id')->references('id')->on('orders')->onDelete('cascade');
            $table->integer('product_id')->unsigned();
            $table->foreign('product_id')->references('id')->on('products');
            $table->integer('amount');
            $table->timestamps();
        });
    }
```
I had to use composite keys to actually get to the "add item" page if not I got this error(snippet below) which suggest there is either something I did wrong or an error in the way the parts get selected due to not properly formatting the SQL select. I could have left the pivot table as normal and looked into the code, but a composite key at least got me further ahead.
> SQLSTATE[23000]: Integrity constraint violation: 1052 Column 'id' in field list is ambiguous (SQL: select id from products inner join  order_product on products.id = order_product.product_id where order_product.order_id is null) (View: D:\WindowsFolders\Documents\PHP\Voyager-test\vendor\tcg\voyager\resources\views\bread\edit-add.blade.php)

Here is the child table:
```PHP
public function up()
    {
        Schema::create('products', function (Blueprint $table) {
            $table->increments('id');
            $table->string('title')->nullable();
            $table->text('description');
            $table->string('image');
            $table->integer('price');
            $table->string('slug')->unique();
            $table->timestamps();
        });
    }
```
Here is the Model function:
```PHP
public function products()
    {
        return $this->belongsToMany(Product::class, 'order_product');
    }
```

Here is just a few dummy products.
![Image of products](https://puu.sh/u7kc0/59a75d2184.png)
They don't seem to show up properly when I select them in the products menu. 
![Image of add page](https://puu.sh/u7kag/c95e7f3ba6.png)
The "Order" BREAD:
![image of bread](https://i.stack.imgur.com/Gi5hK.png)

I'm no real programmer yet and can't really help fix anything without great struggle, but I have a question about the many to many relationship. Do you guys plan to expand on it in the future? Because from what it seems like, it's a little restrictive. How would I change some extra data in the pivot table like the "amount" field I added into my pivot table. Because then I could easily define the amount of products of this type is related to the order.

Django admin have a very nice relationship system where you simply leave your models "as is" with no extra field to store data. And just add the relationship in an admin config. Then the admin page simply add the necessary inputs like in the picture below and you could add many children at the same time. The children wouldn't have to be premade. It might be too much to ask for a django like relationship system.
![Image of django admin](https://djangounchain.files.wordpress.com/2013/02/screen-shot-2013-02-16-at-11-42-52-pm.png)

**Comment by abacram** [comment]
I had a similar issue when listing models in a many to many relationship. In my case, I unchecked the "browse" checkbox in the relationship (products in your case) for the bread configuration and it works fine.

Hope it helps!

**Comment by hzung** [comment]
@abacram: It's worked great! Thanks a lot, bro.

**Comment by adriangordon1231** [comment]
What I had to do to get this to work is define the relationship in two function. 

In your case, you would have to implement belongsToMany in certificationsId() AS WELL AS  a certifications_id() function (the latter of which seems to be required in order for the model to save without throwing a 'Call to undefined method ' error).

**Comment by fletch3555** [comment]
@adriangordon1231, that's a workaround, not a fix.  But it hints at the core issue being that we must've missed something when changing from `type_id()` to `typeId()`

**Comment by adriangordon1231** [comment]
@fletch3555 Never said it was a fix. All I did outline what I did to make it work.

**Comment by handiwijoyo** [comment]
Close this due to old version of Voyager. Please open new issue if this haven't been fixed.
